gpu_bdb/bdb_tools/q03_utils.py

```diff
@@ -88,11 +89,49 @@ def find_items_viewed_before_purchase_kernel(
             else:
                 out_col[i * N + k - 1] = 0
 
+@jit(nopython=True)
+def find_items_viewed_before_purchase_kernel_cpu(
+    relevant_idx_col, user_col, timestamp_col, item_col, out_col, N
+):
+    """
+    Find the past N items viewed after a relevant purchase was made,
+    as defined by the configuration of this query.
+    """
+#     i = cuda.grid(1)
+    relevant_item = q03_purchased_item_IN
 
-def apply_find_items_viewed(df, item_mappings):
+    for i in range(relevant_idx_col.size):  # boundary guard
+        # every relevant row gets N rows in the output, so we need to map the indexes
+        # back into their position in the original array
+        orig_idx = relevant_idx_col[i]
+        current_user = user_col[orig_idx]
+
+        # look at the previous N clicks (assume sorted descending)
+        rows_to_check = N
+        remaining_rows = user_col.size - orig_idx
+
+        if remaining_rows <= rows_to_check:
+            rows_to_check = remaining_rows - 1
+
+        for k in range(1, rows_to_check + 1):
+            if current_user != user_col[orig_idx + k]:
+                out_col[i * N + k - 1] = 0
 
+            # only checking relevant purchases via the relevant_idx_col
+            elif (timestamp_col[orig_idx + k] <= timestamp_col[orig_idx]) & (
+                timestamp_col[orig_idx + k]
+                >= (timestamp_col[orig_idx] - q03_days_in_sec_before_purchase)
+            ):
+                out_col[i * N + k - 1] = item_col[orig_idx + k]
+            else:
+                out_col[i * N + k - 1] = 0
```

Comment:
All of this logic is mostly repeated, we should put this in a function rather than duplicating the logic.

Reply:
@VibhuJawa Unfortunately, we can't put the repeated code into a helper function. Basically, the repeated code must be compiled separately for @cuda and @jit. So I'm not sure how to do that if we don't have two separate functions like we do now.

Reply:
So below seems to be working for both the cases. Does the code error if you do something like below ?  I would really like not to maintain two copies of the code where ever possible. 

```python
@jit(nopython=True)
def common_cpu_gpu_code(out_col,i):
    out_col[i] = i*2
    
@jit(nopython=True)
def cpu_sample_kernel(len_col,out_col):
    for i in range(len_col):
        ## repeated code across CPU and GPU functions
        common_cpu_gpu_code(out_col,i)
        
    return None

@cuda.jit()
def gpu_sample_kernel(len_col,out_col):
    i = cuda.grid(1)
    if i<len_col:    
        ## repeated code across CPU and GPU functions
        common_cpu_gpu_code(out_col,i)
    return None


sample_ar_cpu = np.zeros(100, dtype=np.int32)
cpu_sample_kernel(len(sample_ar_cpu), sample_ar_cpu)
print(sample_ar_cpu)


sample_ar_gpu = cp.zeros(100, dtype=np.int32)
gpu_sample_kernel.forall(len(sample_ar_gpu))(len(sample_ar_gpu), sample_ar_gpu)
print(sample_ar_gpu)
```